// @lifeart/ember-language-server/lib/definition-providers/entry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const file_extension_1 = require("./../utils/file-extension");
const template_1 = require("./template");
const script_1 = require("./script");
const glimmer_script_1 = require("./glimmer-script");
class DefinitionProvider {
    constructor(server) {
        this.server = server;
        this.template = new template_1.default(server);
        this.script = new script_1.default(server);
        this.glimmerScript = new glimmer_script_1.default(server);
    }
    async handle(params) {
        const uri = params.textDocument.uri;
        // this.server.setStatusText('Running');
        const project = this.server.projectRoots.projectForUri(uri);
        if (!project) {
            return null;
        }
        try {
            const extension = (0, file_extension_1.getExtension)(params.textDocument);
            if (extension === '.hbs') {
                return await this.template.handle(params, project);
            }
            else if (extension === '.js' || extension === '.ts') {
                return await this.script.handle(params, project);
            }
            else if (extension === '.gts' || extension === '.gjs') {
                return await this.glimmerScript.handle(params, project);
            }
            else {
                return null;
            }
        }
        catch (e) {
            // logError(e);
            return null;
        }
    }
    get handler() {
        return this.handle.bind(this);
    }
}
exports.default = DefinitionProvider;
