// @lifeart/ember-language-server/lib/definition-providers/script.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const glimmer_utils_1 = require("./../glimmer-utils");
const ember_meta_explorer_1 = require("ember-meta-explorer");
const estree_utils_1 = require("./../estree-utils");
const addon_api_1 = require("./../utils/addon-api");
class ScriptDefinitionProvider {
    constructor(server) {
        this.server = server;
    }
    async handle(params, project) {
        const uri = params.textDocument.uri;
        const { root } = project;
        const document = this.server.documents.get(uri);
        if (!document) {
            return null;
        }
        const content = document.getText();
        const ast = (0, ember_meta_explorer_1.parseScriptFile)(content, {
            sourceType: 'module',
        });
        const astPath = glimmer_utils_1.default.toPosition(ast, (0, estree_utils_1.toPosition)(params.position), content);
        if (!astPath) {
            return null;
        }
        const results = await (0, addon_api_1.queryELSAddonsAPIChain)(project.builtinProviders.definitionProviders, root, {
            focusPath: astPath,
            type: 'script',
            textDocument: params.textDocument,
            position: params.position,
            results: [],
            server: this.server,
        });
        const addonResults = await (0, addon_api_1.queryELSAddonsAPIChain)(project.providers.definitionProviders, root, {
            focusPath: astPath,
            type: 'script',
            textDocument: params.textDocument,
            position: params.position,
            results,
            server: this.server,
        });
        return addonResults;
    }
}
exports.default = ScriptDefinitionProvider;
