// @lifeart/ember-language-server/lib/estree-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.containsPosition = exports.newLocation = exports.toLSRange = exports.toLSPosition = exports.toPosition = exports.comparePositions = exports.newPosition = void 0;
const node_1 = require("vscode-languageserver/node");
function newPosition(line, column) {
    return { line, column };
}
exports.newPosition = newPosition;
function comparePositions(a, b) {
    if (a.line < b.line)
        return -1;
    if (a.line > b.line)
        return 1;
    if (a.column < b.column)
        return -1;
    if (a.column > b.column)
        return 1;
    return 0;
}
exports.comparePositions = comparePositions;
function toPosition(lsp) {
    return newPosition(lsp.line + 1, lsp.character);
}
exports.toPosition = toPosition;
function toLSPosition(pos) {
    return node_1.Position.create(pos.line - 1, pos.column);
}
exports.toLSPosition = toLSPosition;
function toLSRange(loc) {
    return node_1.Range.create(toLSPosition(loc.start), toLSPosition(loc.end));
}
exports.toLSRange = toLSRange;
function newLocation(startLine, startColumn, endLine, endColumn) {
    const start = { line: startLine, column: startColumn };
    const end = { line: endLine, column: endColumn };
    return { start, end };
}
exports.newLocation = newLocation;
function containsPosition(loc, position) {
    return comparePositions(position, loc.start) >= 0 && comparePositions(position, loc.end) <= 0;
}
exports.containsPosition = containsPosition;
