// @lifeart/ember-language-server/lib/builtin-addons/doc/autocomplete.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.docForAttribute = exports.argumentsForBuiltinComponent = exports.valuesForBuiltinComponentArgument = void 0;
const attributes = {
    '...attributes': {
        documentation: `
        In general, you should place ...attributes after any attributes you specify to give people using your component an opportunity to override your attribute. 
        If ...attributes appears after an attribute, it overrides that attribute. 
        If it appears before an attribute, it does not.
        Place ...attributes before your attributes only if you want to disallow tags from overriding your attributes.
        This is likely to be unusual.
        In addition, the class attribute is special, and will be merged with any existing classes on the element rather than overwriting them.
        This allows you to progressively add CSS classes to your components, and makes them more flexible overall.
        `,
    },
};
const builtins = {
    each: {
        arguments: {
            key: {
                documentation: `
            The \`key\` option is used to tell Ember how to determine if the items in the
            array being iterated over with \`{{#each}}\` has changed between renders. By
            default the item's object identity is used.
        `,
                values: [
                    {
                        value: '@identity',
                        documentation: 'The index of the item in the array.',
                    },
                    {
                        value: '@index',
                        documentation: 'The item in the array itself.',
                    },
                ],
            },
        },
    },
};
function valuesForBuiltinComponentArgument(componentName, argumentName) {
    const component = builtins[componentName];
    if (!component) {
        return [];
    }
    const argument = component.arguments[argumentName];
    if (!argument) {
        return [];
    }
    return argument.values.map((e) => {
        return {
            label: e.value,
            documentation: normalizeNewlines(e.documentation),
        };
    });
}
exports.valuesForBuiltinComponentArgument = valuesForBuiltinComponentArgument;
function argumentsForBuiltinComponent(componentName) {
    if (builtins[componentName] && builtins[componentName].arguments) {
        return Object.keys(builtins[componentName].arguments).map((key) => {
            var _a;
            return {
                label: key,
                documentation: normalizeNewlines((_a = builtins[componentName].arguments[key].documentation) !== null && _a !== void 0 ? _a : ''),
            };
        });
    }
    return [];
}
exports.argumentsForBuiltinComponent = argumentsForBuiltinComponent;
function normalizeNewlines(text) {
    return text
        .split('\n')
        .map((e) => e.trim())
        .join('\n');
}
function docForAttribute(name) {
    if (name in attributes) {
        if (attributes[name].documentation) {
            return normalizeNewlines(attributes[name].documentation);
        }
    }
    return '';
}
exports.docForAttribute = docForAttribute;
