// @lifeart/ember-language-server/lib/code-action-provider/entry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeActionProvider = void 0;
const addon_api_1 = require("./../utils/addon-api");
class CodeActionProvider {
    constructor(server) {
        this.server = server;
    }
    async provideCodeActions({ textDocument, context, range }) {
        const project = this.server.projectRoots.projectForUri(textDocument.uri);
        const document = this.server.documents.get(textDocument.uri);
        if (!project || !document) {
            return [];
        }
        const internalResults = await (0, addon_api_1.queryELSAddonsAPIChain)(project.builtinProviders.codeActionProviders, project.root, {
            textDocument,
            context,
            range,
            results: [],
            project: project,
            document: document,
            server: this.server,
        });
        const addonResults = await (0, addon_api_1.queryELSAddonsAPIChain)(project.providers.codeActionProviders, project.root, {
            textDocument,
            context,
            range,
            results: [],
            project: project,
            document: document,
            server: this.server,
        });
        return [...internalResults, ...addonResults].filter(Boolean);
    }
}
exports.CodeActionProvider = CodeActionProvider;
