// @lifeart/ember-language-server/lib/folding-provider/template-folding-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vscode_languageserver_1 = require("vscode-languageserver");
const syntax_1 = require("@glimmer/syntax");
class TemplateFoldingProvider {
    handle(document) {
        const content = document.getText();
        const ast = (0, syntax_1.preprocess)(content);
        const results = [];
        function nodeToResults(node) {
            const loc = node.loc.toJSON();
            if (loc.start.line !== loc.end.line) {
                const kind = node.type === 'MustacheCommentStatement' ? vscode_languageserver_1.FoldingRangeKind.Comment : vscode_languageserver_1.FoldingRangeKind.Region;
                results.push(vscode_languageserver_1.FoldingRange.create(loc.start.line - 1, loc.end.line - 1, loc.start.column, loc.end.column, kind));
            }
        }
        (0, syntax_1.traverse)(ast, {
            BlockStatement(node) {
                nodeToResults(node);
                if (node.inverse) {
                    nodeToResults(node.inverse);
                }
            },
            MustacheCommentStatement(node) {
                nodeToResults(node);
            },
            ElementNode(node) {
                nodeToResults(node);
            },
        });
        return results;
    }
}
exports.default = TemplateFoldingProvider;
