// @lifeart/ember-language-server/lib/utils/fs-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertToFsStat = exports.fileTypeFromFsStat = exports.FileType = void 0;
var FileType;
(function (FileType) {
    FileType[FileType["Directory"] = 2] = "Directory";
    FileType[FileType["File"] = 1] = "File";
    FileType[FileType["SymbolicLink"] = 64] = "SymbolicLink";
    FileType[FileType["Unknown"] = 0] = "Unknown";
})(FileType = exports.FileType || (exports.FileType = {}));
function fileTypeFromFsStat(item) {
    let fType = FileType.Unknown;
    if (item.isDirectory()) {
        fType = FileType.Directory;
    }
    else if (item.isSymbolicLink()) {
        fType = FileType.SymbolicLink;
    }
    else if (item.isFile()) {
        fType = FileType.File;
    }
    return fType;
}
exports.fileTypeFromFsStat = fileTypeFromFsStat;
function convertToFsStat(item) {
    const data = {
        isFile() {
            return item.type === FileType.File;
        },
        isDirectory() {
            return item.type === FileType.Directory;
        },
        isSymbolicLink() {
            return item.type === FileType.SymbolicLink;
        },
        isBlockDevice() {
            return false;
        },
        isCharacterDevice() {
            return false;
        },
        isSocket() {
            return false;
        },
        isFIFO() {
            return false;
        },
        dev: 1,
        ino: 1,
        mode: 1,
        nlink: 1,
        uid: 1,
        gid: 1,
        rdev: 1,
        size: item.size,
        blksize: 1,
        blocks: 1,
        atimeMs: item.ctime,
        mtimeMs: item.mtime,
        ctimeMs: item.ctime,
        birthtimeMs: 1,
        get atime() {
            return new Date(this.atimeMs);
        },
        get mtime() {
            return new Date(this.mtimeMs);
        },
        get ctime() {
            return new Date(this.ctimeMs);
        },
        get birthtime() {
            return new Date(this.birthtimeMs);
        },
    };
    return data;
}
exports.convertToFsStat = convertToFsStat;
